src/main/java/com/n1analytics/paillier/PaillierContext.java

```diff
@@ -431,222 +354,296 @@ public boolean isValid(EncodedNumber encoded) {
-   * the {@code Number} representation of the {@code double} to be encoded is not valid.
+   * Decreases the exponent of an {@code EncodedNumber} to {@code newExp}. If {@code newExp} is greater than
+   * the {@code EncodedNumber}'s current {@code exponent}, throws an IllegalArgumentException.
    *
-   * @param value the {@code double} to be encoded.
-   * @return the encoding result.
-   * @throws EncodeException if the {@code value} is not valid.
+   * @param encodedNumber the {@code EncodedNumber} which {@code exponent} will be reduced.
+   * @param newExp the new {@code exponent}, must be less than the current {@code exponent}.
+   * @return an {@code EncodedNumber} representing the same value with {@code exponent} equals to {@code newExp}.
    */
-  public EncodedNumber encode(double value) throws EncodeException {
-    return encode(Number.encode(value));
+  public EncodedNumber decreaseExponentTo(EncodedNumber encodedNumber, int newExp) {
+    BigInteger significand = encodedNumber.getValue();
+    int exponent = encodedNumber.getExponent();
+    if(newExp > exponent){
+      throw new IllegalArgumentException("New exponent: "+ newExp +
+              "should be more negative than old exponent: " + exponent + ".");
+    }
+
+    int expDiff = exponent - newExp;
+    BigInteger bigFactor = getRescalingFactor(expDiff);
+    BigInteger newEnc = significand.multiply(bigFactor).mod(publicKey.getModulus());
+    return new EncodedNumber(this, newEnc, newExp);
   }
 
   /**
-   * Encodes a {@code long} using this {@code PaillierContext}. Throws an EncodeException if
-   * the {@code Number} representation of the {@code long} to be encoded is not valid.
+   * Decreases the exponent of an {@code EncryptedNumber} to {@code newExp}. If {@code newExp} is greater than
+   * the {@code EncryptedNumber}'s current {@code exponent}, throws an IllegalArgumentException.
    *
-   * @param value the {@code long} to be encoded.
-   * @return the encoding result.
-   * @throws EncodeException if the {@code value} is not valid.
+   * @param encryptedNumber the {@code EncryptedNumber} which {@code exponent} will be reduced.
+   * @param newExp the new {@code exponent}, must be less than the current {@code exponent}.
+   * @return an {@code EncryptedNumber} representing the same value with {@code exponent} equals to {@code newExp}.
    */
-  public EncodedNumber encode(long value) throws EncodeException {
-    return encode(Number.encode(value));
+  public EncryptedNumber decreaseExponentTo(EncryptedNumber encryptedNumber, int newExp) {
+    int exponent = encryptedNumber.getExponent();
+    if(newExp > exponent){
+      throw new IllegalArgumentException("New exponent: "+ newExp +
+              "should be more negative than old exponent: " + exponent + ".");
+    }
+
+    int expDiff = exponent - newExp;
+    BigInteger bigFactor = getRescalingFactor(expDiff);
+    BigInteger newEnc = publicKey.raw_multiply(encryptedNumber.calculateCiphertext(), bigFactor);
```

Comment:
better: use 
`BigInteger newEnc = publicKey.raw_multiply(encryptedNumber.ciphertext, bigFactor);`
`return new EncryptedNumber(this, newEnc, newExp, encryptedNumber.isSafe);`
Otherwise you might do one unnecessary obfuscation.